### PR TITLE
Feature/prioritize transactions

### PR DESCRIPTION
**Pull request by comountainclimber**
screen shot 2018-06-09 at 2 38 05 pm

**What current issue(s) from Trello/Github does this address?**

https://github.com/CityOfZion/neon-wallet/issues/1202

**What problem does this PR solve?**
This allows for users to prioritize transactions using existing parts of the neon-js API. We can port this over to v2 after merging... I know this has been a priority (no pun intended) for a while now

**How did you solve this problem?**

**How did you make sure your solution works?**
Tested locally and verified via https://neotracker.io/tx/7de63fb646827b9cd261b0b0b253b1035882a97473fd22887c73979090852daa

**Are there any special changes in the code that we should be aware of?**

**Is there anything else we should know?**
If more gas is added to the fee than is available in balance an alert is popped.
screen shot 2018-06-09 at 2 39 34 pm

This upgrades neon-js dependency to 3.8.0.

- [ ] Unit tests written?

### CODE REVIEW

File: app/modules/transactions.js [hunk 1]
```diff
@@ -77,14 +77,20 @@ const makeRequest = (sendEntries: Array<SendEntryType>, config: Object) => {
...
+
   if (script === '') {
     return api.sendAsset({ ...config, intents: buildIntents(sendEntries) }, api.neoscan)
   } else {
     return api.doInvoke({
       ...config,
       intents: buildIntents(sendEntries),
       script,
-      gas: 0
+      gas: priorityFees
```
Comment (comountainclimber) [inline app/modules/transactions.js]: @snowypowers if you could take a look specifically here (Im pretty sure this is incorrect as it is not working currently) and explain where I need to attach priority fees this PR will be 👍

Comment (snowypowers) [reply app/modules/transactions.js]:
```

Comment (comountainclimber) [reply app/modules/transactions.js]: Thanks!

**Comment by comountainclimber** [comment]
Ahhh ok thank you very much @snowypowers I will go ahead and refactor the UI

**Comment by comountainclimber** [comment]
The UI has been updated and manual testing is complete

**Comment by comountainclimber** [comment]
@snowypowers I addressed your comment about UI/UX and refactored PR... is there anything else you think I need to change??

File: app/components/Modals/SendModal/ConfirmDisplay.jsx [hunk 2]
```diff
@@ -55,9 +61,28 @@ export default class ConfirmDisplay extends React.Component<Props, State> {
...
+          <div className={addRecipientDisplayStyles.row}>
+            <div id='sendAmount' className={addRecipientDisplayStyles.column}>
+              <label className={addRecipientDisplayStyles.label}>Priority Fee:</label>
+              <NumberInput
+                value={priorityFee}
+                placeholder='Priority Fee'
+                options={{ numeralDecimalScale: COIN_DECIMAL_LENGTH }}
+                onChange={onUpdatePriorityFee} />
+              <label className={addRecipientDisplayStyles.label}>
+                ({balances['GAS']} GAS Available)
```
Comment (evgenyboxer) [inline app/components/Modals/SendModal/ConfirmDisplay.jsx]: Use `balances[ASSETS.GAS]`

File: app/components/Modals/SendModal/ConfirmDisplay.jsx [hunk 3]
```diff
@@ -3,33 +3,39 @@ import React from 'react'
...
 import { Address } from '../../Blockchain'
-import { formatBalance } from '../../../core/formatters'
+import { formatBalance, COIN_DECIMAL_LENGTH } from '../../../core/formatters'
 
 import styles from './ConfirmDisplay.scss'
+import addRecipientDisplayStyles from './AddRecipientDisplay.scss'
 
 type Props = {
   address: string,
+  balances: Object,
```
Comment (evgenyboxer) [inline app/components/Modals/SendModal/ConfirmDisplay.jsx]: Use `BalancesType`

File: app/components/Modals/SendModal/SendModal.jsx [hunk 4]
```diff
@@ -143,9 +149,14 @@ export default class SendModal extends Component<Props, State> {
   }
 
   handleConfirmTransaction = () => {
-    const { sendTransaction } = this.props
-    const { entries } = this.state
-    sendTransaction(entries).then(this.close)
+    const { showErrorNotification, sendTransaction } = this.props
+    const { priorityFee, balances, entries } = this.state
+
+    if (priorityFee && toBigNumber(priorityFee).gt(balances.GAS)) {
```
Comment (evgenyboxer) [inline app/components/Modals/SendModal/SendModal.jsx]: `balances[ASSETS.GAS]`

File: app/modules/transactions.js [hunk 5]
```diff
@@ -132,6 +131,7 @@ export const sendTransaction = (sendEntries: Array<SendEntryType>) => async (
       tokensBalanceMap,
       address: fromAddress,
       publicKey,
+      fees: Number(priorityFee),
```
Comment (evgenyboxer) [inline app/modules/transactions.js]: Use `toNumber` from `core/math.js`

Comment (evgenyboxer) [reply app/modules/transactions.js]: Also, where is `fees` used?

**Comment by evgenyboxer** [comment]
UI Wise, what happens if they got no GAS at all?

File: app/modules/transactions.js (hunk 5 above)
Comment (snowypowers) [reply app/modules/transactions.js]: `fees` is passed as part of the config object into `sendAsset` / `doInvoke` which will handle the rest.

**Comment by snowypowers** [comment]
My approval is for interactions with the backend. I will let evgeny handle the rest of the front end review

**Comment by comountainclimber** [comment]
@evgenyboxer If they have no gas the alert is shown. See my description of PR

**Comment by evgenyboxer** [comment]
@comountainclimber could you add the fee priority as a toggle?
"Add priority fee" - that would trigger the label+input...  (instead of a blank text input) as this isn't something that everyone would use all the time.

**Comment by comountainclimber** [comment]
@evgenyboxer @drptbl Thank you guys for the excellent feedback and for testing! I will go ahead and implement your suggestions. In terms of @drptbl comments for item #3 was planning on implementing the prompt if priority fee is greater than 0.1 GAS. If someone thinks the threshold should be greater or lower let me know

**Comment by comountainclimber** [comment]
The UI/UX has been updated as per the suggestions... I also polished up the modal sizes:
screen shot 2018-06-11 at 10 21 39 pm
screen shot 2018-06-11 at 10 21 44 pm

Theres still no alert if you enter in more than .1 GAS let me know if you think that is still necessary @evgenyboxer

File: app/components/Modals/SendModal/SendModal.jsx [hunk 6]
```diff
@@ -85,15 +89,27 @@ export default class SendModal extends Component<Props, State> {
...
   handleAddRecipient = () => {
     this.setState({ display: DISPLAY_MODES.ADD_RECIPIENT })
   }
 
+  determineModalHeight = () => {
+    const { display, priorityFeeCollapsed } = this.state
+    if (display === DISPLAY_MODES.ADD_RECIPIENT) {
+      return '350px'
+    } else if (priorityFeeCollapsed) return '350px'
+    return '485px'
```
Comment (evgenyboxer) [inline app/components/Modals/SendModal/SendModal.jsx]: shouldn't this be: 

`return display === DISPLAY_MODES.ADD_RECIPIENT || priorityFeeCollapsed ? '350px' : '485px'` ? (you return the same value in 2 conditions)

also, would be best if these are constants

**Comment by comountainclimber** [comment]
@evgenyboxer added the constants to core/constants and fixed the weird function... let me know if there is anything else and thanks for the review 👍

**Comment by evgenyboxer** [comment]
@drptbl you can QA this final piece, and we can finally merge 👍

**Comment by drptbl** [comment]
@comountainclimber @evgenyboxer 
There is a new bug which has to be fixed before this is merged. After adding a fee, changing it to any amount, then removing and sending transaction - fee is still added.

Reproduction steps:
1. Login to the wallet
2. Click on "Send" button
3. Add any amount of any token to transaction
4. Click on "Next" button
5. Click on "Add Priority Fee" button
6. Fill any amount of priority fee (or click on "Add smallest fee possible")
7. Click on "Remove Priority Fee" button
8. Click on "I agree~" checkbox
9. Click on "Send Assets" button

After transaction is sent, previous fee which was removed by "Remove Priority Fee" button is still added to the transaction.

**Comment by comountainclimber** [comment]
Weird I thought I manually tested this scenario... Anyway excellent catch @drptbl 🙏 I will go ahead and address this later today.

**Comment by comountainclimber** [comment]
@drptbl should be good to go now

**Comment by drptbl** [comment]
@evgenyboxer @comountainclimber Looks good now 👍.

**Pull request merged**